sympy/polys/matrices/domainmatrix.py

```diff
@@ -357,6 +478,11 @@ def hstack(A, B):
         unify
 
         """
+
+        if isinstance(A.rep, list) != isinstance(B.rep, list):
+            A = A.to_dense()
+            B = B.to_dense()
+
         A, B = A.unify(B)
```

Comment:
The logic for combining dense and sparse could be handled in `unify`. Perhaps there could be an option to specify whether to unify towards sparse or dense if they are mixed or whether to always convert to dense or sparse or something like that. I think in basic terms `A.unify(B)` should make the two matrices compatible for subsequent operations.

Reply:
There could also be separate functions like `A.unify_domain` and `A.unify_fmt` and then general `A.unify` could call both.

Reply:
Maybe operations like `a + b` should be changed to unify the domains rather than raising an exception.

Reply:
Ok, I'll make those changes. 
What do you think of converting both matrices to dense if their `rep` doesn't match. Should there be an option to convert both to sparse?

Reply:
I'm not entirely sure whether everything should unify automatically or whether explicit conversion should be done by the user of `DomainMatrix`. Still thinking about it...

For now to finish this you could just make it so that mixing dense and sparse is an error. Then perhaps later if we want automatic conversions we can add them.

Reply:
Well I did unify automatically, for binary operations in latest commit. I saw this message after pushing that change